UefiTestingPkg/AuditTests/PagingAudit/UEFI/Dxe/App/DxePagingAuditTestApp.c

```diff
@@ -29,7 +163,129 @@ DxePagingAuditTestAppEntryPoint (
   IN     EFI_SYSTEM_TABLE  *SystemTable
   )
 {
-  DumpPagingInfo (NULL, NULL);
+  EFI_STATUS                     Status;
+  UNIT_TEST_FRAMEWORK_HANDLE     Fw   = NULL;
+  UNIT_TEST_SUITE_HANDLE         Misc = NULL;
+  PAGING_AUDIT_TEST_CONTEXT      *Context;
+  IA32_CR4                       Cr4;
+  PAGING_MODE                    PagingMode;
+  IA32_MAP_ENTRY                 *Map           = NULL;
+  UINTN                          MapCount       = 0;
+  UINTN                          PagesAllocated = 0;
+  BOOLEAN                        RunTests       = TRUE;
+  EFI_SHELL_PARAMETERS_PROTOCOL  *ShellParams;
+
+  DEBUG ((DEBUG_ERROR, "%a()\n", __FUNCTION__));
+
+  DEBUG ((DEBUG_ERROR, "%a v%a\n", UNIT_TEST_APP_NAME, UNIT_TEST_APP_VERSION));
+
+  Status = gBS->HandleProtocol (
+                  gImageHandle,
+                  &gEfiShellParametersProtocolGuid,
+                  (VOID **)&ShellParams
+                  );
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "%a Could not retrieve command line args!\n", __FUNCTION__));
+    return EFI_PROTOCOL_ERROR;
+  }
+
+  if (ShellParams->Argc > 1) {
+    RunTests = FALSE;
+    if (StrnCmp (ShellParams->Argv[1], L"-r", 4) == 0) {
+      RunTests = TRUE;
+    } else if (StrnCmp (ShellParams->Argv[1], L"-d", 4) == 0) {
+      DumpPagingInfo (NULL, NULL);
+    } else {
+      if (StrnCmp (ShellParams->Argv[1], L"-h", 4) != 0) {
+        DEBUG ((DEBUG_INFO, "Invalid argument. Use \'-h\' to see a list of valid arguments.\n"));
+      }
+
+      DEBUG ((DEBUG_INFO, "-h : Print available flags\n"));
+      DEBUG ((DEBUG_INFO, "-d : Dump the page table files to the EFI partition\n"));
+      DEBUG ((DEBUG_INFO, "-r : Run the application tests\n"));
+      DEBUG ((DEBUG_INFO, "NOTE: Combined flags (i.e. -rd) is not supported\n"));
+    }
+  }
+
+  if (RunTests) {
+    Context = (PAGING_AUDIT_TEST_CONTEXT *)AllocateZeroPool (sizeof (PAGING_AUDIT_TEST_CONTEXT));
+
+    if (Context == NULL) {
+      DEBUG ((DEBUG_ERROR, "Failed to allocate test context\n"));
+      goto EXIT;
+    }
+
+    //
+    // Start setting up the test framework for running the tests.
+    //
+    Status = InitUnitTestFramework (&Fw, UNIT_TEST_APP_NAME, gEfiCallerBaseName, UNIT_TEST_APP_VERSION);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "Failed in InitUnitTestFramework. Status = %r\n", Status));
+      goto EXIT;
+    }
+
+    // Poll CR4 to deterimine the page table depth
+    Cr4.UintN = AsmReadCr4 ();
+
+    if (Cr4.Bits.LA57 != 0) {
+      PagingMode = Paging5Level;
+    } else {
+      PagingMode = Paging4Level;
+    }
+

```

Comment:
CR3 is X86 specific. So this app cannot be used on any other architectures?

Reply:
Correct. Even the common portion of the app uses x86 opinionated logic such as x86 page table definitions, so it will need to be broken into x86 and ARM versions in the future.